### PR TITLE
Resolving Kamt issue#1862

### PR DESCRIPTION
**Pull request by lordshashank**
- Resolving https://github.com/filecoin-project/ref-fvm/issues/1862
- implemented iter and iter_from for kamt referring to hamt implementation.
- wrote simple test for iter_from [here](https://github.com/lordshashank/ref-fvm/blob/088dcbf015756ac7bd887cfa273d2de522b572a4/ipld/kamt/src/kamt.rs#L409)
- Also implemented function to declare kamt with custom bit width [here](https://github.com/lordshashank/ref-fvm/blob/088dcbf015756ac7bd887cfa273d2de522b572a4/ipld/kamt/src/kamt.rs#L86) and wrote its test [here](https://github.com/lordshashank/ref-fvm/blob/088dcbf015756ac7bd887cfa273d2de522b572a4/ipld/kamt/src/kamt.rs#L437)

### CODE REVIEW

**Comment by Stebalien** [comment]
This is almost correct except for `iter_from`. Unlike `HAMT`, the `KAMT` supports "extensions" to collapse keys with spans of identical bits.

You can see how we use them here:

https://github.com/filecoin-project/ref-fvm/blob/088dcbf015756ac7bd887cfa273d2de522b572a4/ipld/kamt/src/node.rs#L241-L244

I'm pretty sure all you need to do is save the extension instead of discarding it, and call:

```rust
        match match_extension(conf, hashed_key, ext)? {
            ExtensionMatch::Full { .. } => {},
            ExtensionMatch::Partial { .. } => return Err(Error::StartKeyNotFound),
        }
```

At the end of every loop. But you'll also need to write a test or two. I.e., make a kamt with the keys 0x00000000000001, 0x00000000000002, 0x10000000000001, 0x10000000000002, 0x20000000000001, 0x20000000000002, etc... then try to start at `0x10000000000001`.

**Comment by Stebalien** [comment]
And sorry this took so long for me to get to!

File: ipld/kamt/src/iter.rs [hunk 1]
```diff
@@ -145,3 +155,40 @@ where
     BS: Blockstore,
 {
 }
+
+/// Helper method to check if a key matches an extension (if there is one)
+/// and return the number of levels skipped. If the key doesn't match,
+/// this will be the number of levels where the extension has to be split.
+fn match_extension<'a>(
```
Comment (Stebalien) [inline ipld/kamt/src/iter.rs:162]: We probably need to export this from the `node` module with `pub(crate)`.

File: ipld/kamt/src/kamt.rs [hunk 2]
```diff
@@ -81,6 +82,17 @@ where
         }
     }
 
+    /// Construct kamt with a bit width
+    pub fn new_with_bit_width(store: BS, bit_width: u32) -> Self {
```
Comment (Stebalien) [inline ipld/kamt/src/kamt.rs:86]: Let's drop this for now. We have it on the HAMT for historical reasons, but `new_with_config` is more general.

File: ipld/kamt/src/kamt.rs [hunk 3]
```diff
@@ -352,6 +364,133 @@ where
...
+    ///     x = x+1;
+    /// }
+    /// assert_eq!(x,2)
+    /// ```
+    pub fn iter(&self) -> IterImpl<BS, V, K, H, N> {
+        IterImpl::new(&self.store, &self.root)
+    }
+
+    /// Iterate over the KAMT starting at the given key.
+    /// iteration:
```
Comment (Stebalien) [inline ipld/kamt/src/kamt.rs:410]: stray line?

File: ipld/kamt/src/kamt.rs [hunk 4]
```diff
@@ -352,6 +364,133 @@ where
...
+    /// Iterate over the KAMT starting at the given key. This can be used to implement "ranged" iteration:
+    ///
+    /// ```rust
+    /// use fvm_ipld_kamt::Kamt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    /// use fvm_ipld_kamt::id::Identity;
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// // Create a Kamt with 5 keys, a-e.
+    /// let mut kamt: Kamt<_, u32, String, Identity> = Kamt::new_with_bit_width(store, 5);
```
Comment (Stebalien) [inline ipld/kamt/src/kamt.rs:446]: I'd just change this to `Kamt::new_with_config` (or drop the bit-width, I don't think that it's actually necessary to specify it?

File: ipld/kamt/src/iter.rs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+use crate::pointer::Pointer;
+use crate::{AsHashedKey, Config, Error, KeyValuePair};
+#[doc(hidden)]
+pub struct IterImpl<'a, BS, V, K, H, const N: usize = 32> {
+    store: &'a BS,
+    stack: Vec<std::slice::Iter<'a, Pointer<K, V, H, N>>>,
+    current: std::slice::Iter<'a, KeyValuePair<K, V>>,
+}
+
+pub type Iter<'a, BS, V, K, H, const N: usize = 32> = IterImpl<'a, BS, V, K, H, N>;
```
Comment (Stebalien) [inline ipld/kamt/src/iter.rs:21]: We don't need a separate `Iter` and `IterImpl`. We only have that for the `Hamt` to handle the multiple versions.

File: ipld/kamt/src/iter.rs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        root: &'a Node<K, V, H, N>,
+        key: &Q,
+        conf: &'a Config,
+    ) -> Result<Self, Error>
+    where
+        K: Borrow<Q>,
+        Q: PartialEq,
+        H: AsHashedKey<Q, N>,
+    {
+        println!("IterImpl::new_from");
```
Comment (Stebalien) [inline ipld/kamt/src/iter.rs:47]: remove println.

File: ipld/kamt/src/iter.rs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        let hashed_key = H::as_hashed_key(key);
+        let mut hash = HashBits::new(&hashed_key);
+        let mut node = root;
+        let mut ext;
```
Comment (Stebalien) [inline ipld/kamt/src/iter.rs:51]: This should be inside the loop (doesn't cost anything extra and keeps us from keeping it around between iterations).

Comment (Stebalien) [reply ipld/kamt/src/iter.rs:51]: And, inside the loop, it doesn't need to be `mut`.

**Review by Stebalien — COMMENTED** [review]
Getting down to the final nits. The only real remaining thing is a good test-case.

File: ipld/kamt/src/iter.rs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::de::DeserializeOwned;
+use fvm_ipld_encoding::CborStore;
+
+use crate::hash_bits::HashBits;
+use crate::node::{match_extension, ExtensionMatch, Node};
+use crate::pointer::Pointer;
+use crate::{AsHashedKey, Config, Error, KeyValuePair};
+#[doc(hidden)]
+pub struct IterImpl<'a, BS, V, K, H, const N: usize = 32> {
```
Comment (Stebalien) [inline ipld/kamt/src/iter.rs:15]: Let's rename this to `Iter` and remove the `#[doc(hidden)]`. We only did it that way because we had multiple versions.

Comment (lordshashank) [reply ipld/kamt/src/iter.rs]: done

**Pull request merged**